Convert sealed trait to sealed abstract class. Mark all classes Serializable

saucam opened the pull request:
solves #286 
@jdegoes for part 3, do you want to add type bounds to IO methods ?

```final def parWith[E1 >: E, B, C](that: IO[E1, B])(f: (A, B) => C): IO[E1, C]```

to

```final def parWith[E1 >: E, B <: Serializable, C <: Serializable](that: IO[E1, B])(f: (A, B) => C): IO[E1, C]```

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Fiber.scala:
```diff
@@ -21,7 +21,7 @@ package scalaz.zio
  * }}}
  */
-trait Fiber[+E, +A] { self =>
+trait Fiber[+E, +A] extends Serializable { self =>
```
This is one I would not change, because a Fiber represents a running computation. It's possible to make it serializable but only with additional development.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ExitResult.scala:
```diff
@@ -6,7 +6,7 @@ package scalaz.zio
  * due to interruption or runtime error.
  */
-sealed trait ExitResult[+E, +A] { self =>
+sealed abstract class ExitResult[+E, +A] extends Serializable { self =>
```
I'd also extend `Product` together with `Serializable` for reasons [discussed elsewhere](https://typelevel.org/blog/2018/05/09/product-with-serializable.html).

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Clock.scala:
```diff
@@ -4,7 +4,7 @@ package scalaz.zio
 import scala.concurrent.duration._
 
-trait Clock {
+trait Clock extends Serializable {
```
I'd look at what the `Live` clock captures to make sure it is indeed serializable.

saucam replied:
Live captures ```nanotime``` which itself is serializable. Also confirmed by unit test.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -42,7 +42,7 @@ import scala.concurrent.ExecutionContext
  * `App`.
  */
-sealed abstract class IO[+E, +A] { self =>
+sealed abstract class IO[+E, +A] extends Serializable { self =>
```
[With Product](https://typelevel.org/blog/2018/05/09/product-with-serializable.html), here and elsewhere.

saucam replied:
Can make the IO extend Product but the materializations of IO are classes and not case classes hence we need to implement the abstract methods of ```Product``` inside, or make these case classes:

```final class Suspend[E, A] private[IO] (val value: () => IO[E, A]) extends IO[E, A]```

jdegoes replied:
You are absolutely right, there's no need to extend `Product` for `IO`. For other types which are case classes, however, let's try to do that.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Queue.scala:
```diff
@@ -17,7 +17,7 @@ import Queue.internal._
  *    easy ways to improve performance.
  */
-class Queue[A] private (capacity: Int, ref: Ref[State[A]]) {
+class Queue[A] private (capacity: Int, ref: Ref[State[A]]) extends Serializable {
```
In order for this to have any meaning, the `State` and of course `Ref` must be serializable.

jdegoes replied:
We can change `Ref` so it is no longer `AnyVal`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Semaphore.scala:
```diff
@@ -9,7 +9,7 @@ import scalaz.zio.ExitResult.Terminated
 import scala.collection.immutable.{ Queue => IQueue }
 
-final class Semaphore private (private val state: Ref[State]) {
+final class Semaphore private (private val state: Ref[State]) extends Serializable {
```
`Ref` and `State` again.

saucam commented:
Added unit test and added a few comments. 
@jdegoes  thanks for your inputs, please do let me know if the pattern of the tests seems fine, I can add more tests on the same lines

jdegoes commented on core/jvm/src/main/scala/scalaz/zio/App.scala:
```diff
@@ -27,7 +27,7 @@ import scala.concurrent.duration.Duration
 trait App extends RTS {
 
-  sealed trait ExitStatus
+  sealed abstract class ExitStatus
```
`extends Serializable with Product`

jdegoes commented on core/jvm/src/main/scala/scalaz/zio/RTS.scala:
```diff
@@ -111,7 +111,7 @@ trait RTS {
 private object RTS {
 
-  sealed trait RaceState
+  sealed abstract class RaceState
```
`extends Serializable with Product`

jdegoes commented on core/jvm/src/main/scala/scalaz/zio/RTS.scala:
```diff
@@ -1059,7 +1059,7 @@ private object RTS {
   }
 
-  sealed trait FiberStatus[E, A] {
+  sealed abstract class FiberStatus[E, A] {
```
`extends Serializable with Product`

jdegoes commented on core/jvm/src/test/scala/scalaz/zio/SerializableSpec.scala:
```diff
@@ -0,0 +1,104 @@
+
+import java.io._
+
+class SerializableSpec extends AbstractRTSSpec {
```
These tests look awesome!

jdegoes commented on core/jvm/src/test/scala/scalaz/zio/SerializableSpec.scala:
```diff
@@ -0,0 +1,104 @@
+
+class SerializableSpec extends AbstractRTSSpec {
+
+  def serializeToBytes[T](a: T): Array[Byte] = {
```
I might hide these two functions inside `serializeAndBack`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Promise.scala:
```diff
@@ -179,7 +179,7 @@ object Promise {
 
   private[zio] object internal {
-    sealed trait State[E, A]
+    sealed abstract class State[E, A]                             extends Serializable
```
`extends Serializable with Product`

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Ref.scala:
```diff
@@ -16,7 +16,7 @@ import java.util.concurrent.atomic.AtomicReference
  * }}}
  */
-final class Ref[A] private (private val value: AtomicReference[A]) extends AnyVal {
+final class Ref[A] private (private val value: AtomicReference[A]) extends AnyVal with Serializable {
```
Does this work, even though it still extends `AnyVal`?

saucam replied:
Yes oddly it works! if I do not extend from ```Serializable``` then my unit test fails though. It may be because I think primitives are by default serializable, but we still need to extend here for it to work

saucam commented:
@jdegoes thanks for the feedback!! Made more changes, kindly review.

jdegoes commented:
@saucam Don't forget the other ones, too (`Semaphore`, `Clock`, etc.).

saucam commented:
```Clock``` object does not have such methods , do we want to add this as a safeguard because as the library evolves we might have more methods in such objects?
Same with ```FiberStatus``` ,  ```RaceState``` , ```ExitStatus```, ```ExitResult```

Marking ```Queue``` object ```Serializable``` results in 
```
encountered unrecoverable cycle resolving import
 import Queue.internal._
```

We will need to move out the ```internal``` object but I don't understand why it happens!

saucam commented:
Not really, this is compile-time error, not sure but something causing cyclic dependency. 
I have moved it out now though.

saucam commented:
Added the needful

saucam commented:
cc @jdegoes

saucam commented:
Thank you for all your help @jdegoes ! hope I can contribute more as I understand more things here
